### PR TITLE
fix: Assert credentials individually on U2F devices

### PR DESCRIPTION
**Pull request by codingllama**
Fix a somewhat rare error in older authenticator models (reproduced with a FIDO U2F Security Key firmware 4.1.8).

The error is as follows: if certain key handles unknown to the authenticator exist in the allowed credentials set, the authenticator fails the assertion with an "internal error". This PR changes assertions in this specific scenario (U2F key, >1 allowed credential) to run sequentially through each key handle looking for a known credential to use. 

#44912

Changelog: Fixes a rare "internal error" on older U2F authenticators when using tsh

### CODE REVIEW

**Comment by codingllama** [comment]
Tested manually in various scenarios: login with multiple keys, registration with multiple keys, registration with device registered/not registered, login as single key, etc.

A few logs added below.

login with multiple keys:

```
2024-08-08T16:42:20-03:00 DEBU             FIDO2: Using libfido2 for assertion webauthncli/api.go:183
2024-08-08T16:42:20-03:00 DEBU             FIDO2: assertion: passwordless=false, uv=false, 4 allowed credentials webauthncli/fido2.go:168
Tap any security key
2024-08-08T16:42:20-03:00 DEBU             FIDO2: Device ioreg://4294988912: not a FIDO2 device webauthncli/fido2.go:862
2024-08-08T16:42:20-03:00 DEBU             FIDO2: Device ioreg://4294970869: info &libfido2.DeviceInfo{Versions:[]string{"U2F_V2", "FIDO_2_0", "FIDO_2_1_PRE"}, Extensions:[]string{"credProtect", "hmac-secret"}, AAGUID:[]uint8{0xee, 0x88, 0x28, 0x79, 0x72, 0x1c, 0x49, 0x13, 0x97, 0x75, 0x3d, 0xfc, 0xce, 0x97, 0x7, 0x2a}, Options:[]libfido2.Option{libfido2.Option{Name:"rk", Value:"true"}, libfido2.Option{Name:"up", Value:"true"}, libfido2.Option{Name:"plat", Value:"false"}, libfido2.Option{Name:"clientPin", Value:"false"}, libfido2.Option{Name:"credentialMgmtPreview", Value:"true"}}, Protocols:[]uint8{0x2, 0x1}} webauthncli/fido2.go:860
2024-08-08T16:42:20-03:00 DEBU             FIDO2: Device ioreg://4294988912: Using credential 9d6b1b54b0... webauthncli/fido2.go:349
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Got 1 assertions webauthncli/fido2.go:251
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Authenticated: credential ID (b64) = nWsbVLAn7rxUE0byLwViJEAkUIOFBk369qeVctD3oLfBqMw-OV1jMpVbyqORUQouWUNxLAzC6HfQrbWTRnE6JQ, user ID (hex) = , user name = "" webauthncli/fido2.go:259
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Device ioreg://4294988912: callback returned, requiresPIN=false, err=<nil> webauthncli/fido2.go:883
Detected security key tap
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Cancel device ioreg://4294970869, err=<nil> webauthncli/fido2.go:826
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Close device ioreg://4294988912, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Cancel device ioreg://4294988912, err=<nil> webauthncli/fido2.go:826
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Device ioreg://4294970869: callback returned, requiresPIN=false, err=failed to get assertion: keep alive cancel webauthncli/fido2.go:883
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Close device ioreg://4294970869, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:42:23-03:00 DEBU             FIDO2: Device goroutines exited cleanly webauthncli/fido2.go:688
```

tsh mfa add (adding another key):

```
2024-08-08T16:44:24-03:00 DEBU             FIDO2: Using libfido2 for credential creation webauthncli/api.go:230
2024-08-08T16:44:24-03:00 DEBU             FIDO2: registration: resident key=false webauthncli/fido2.go:449
Tap your *new* security key
2024-08-08T16:44:24-03:00 DEBU             FIDO2: Device ioreg://4294988912: not a FIDO2 device webauthncli/fido2.go:862
2024-08-08T16:44:24-03:00 DEBU             FIDO2: Device ioreg://4294970869: info &libfido2.DeviceInfo{Versions:[]string{"U2F_V2", "FIDO_2_0", "FIDO_2_1_PRE"}, Extensions:[]string{"credProtect", "hmac-secret"}, AAGUID:[]uint8{0xee, 0x88, 0x28, 0x79, 0x72, 0x1c, 0x49, 0x13, 0x97, 0x75, 0x3d, 0xfc, 0xce, 0x97, 0x7, 0x2a}, Options:[]libfido2.Option{libfido2.Option{Name:"rk", Value:"true"}, libfido2.Option{Name:"up", Value:"true"}, libfido2.Option{Name:"plat", Value:"false"}, libfido2.Option{Name:"clientPin", Value:"false"}, libfido2.Option{Name:"credentialMgmtPreview", Value:"true"}}, Protocols:[]uint8{0x2, 0x1}} webauthncli/fido2.go:860
2024-08-08T16:44:24-03:00 DEBU             FIDO2: Device ioreg://4294970869: filtered due to presence of excluded credential webauthncli/fido2.go:529
2024-08-08T16:44:24-03:00 DEBU             FIDO2: Device ioreg://4294970869 filtered, err=device already holds a registered credential webauthncli/fido2.go:869
2024-08-08T16:44:27-03:00 DEBU             FIDO2: Device ioreg://4294988912: callback returned, requiresPIN=false, err=<nil> webauthncli/fido2.go:883
Detected security key tap
2024-08-08T16:44:27-03:00 DEBU             FIDO2: Cancel device ioreg://4294970869, err=<nil> webauthncli/fido2.go:826
2024-08-08T16:44:27-03:00 DEBU             FIDO2: Close device ioreg://4294988912, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:44:27-03:00 DEBU             FIDO2: Cancel device ioreg://4294988912, err=<nil> webauthncli/fido2.go:826
2024-08-08T16:44:27-03:00 DEBU             FIDO2: Device touch status error: touch status: keep alive cancel webauthncli/fido2.go:1049
2024-08-08T16:44:27-03:00 DEBU             FIDO2: Close device ioreg://4294970869, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:44:27-03:00 DEBU             FIDO2: Device goroutines exited cleanly webauthncli/fido2.go:688
```

tsh mfa add (already registered):

```
2024-08-08T16:46:14-03:00 DEBU             FIDO2: Using libfido2 for credential creation webauthncli/api.go:230
2024-08-08T16:46:14-03:00 DEBU             FIDO2: registration: resident key=false webauthncli/fido2.go:449
Tap your *new* security key
2024-08-08T16:46:14-03:00 DEBU             FIDO2: Device ioreg://4294988912: not a FIDO2 device webauthncli/fido2.go:862
2024-08-08T16:46:14-03:00 DEBU             FIDO2: Device ioreg://4294970869: info &libfido2.DeviceInfo{Versions:[]string{"U2F_V2", "FIDO_2_0", "FIDO_2_1_PRE"}, Extensions:[]string{"credProtect", "hmac-secret"}, AAGUID:[]uint8{0xee, 0x88, 0x28, 0x79, 0x72, 0x1c, 0x49, 0x13, 0x97, 0x75, 0x3d, 0xfc, 0xce, 0x97, 0x7, 0x2a}, Options:[]libfido2.Option{libfido2.Option{Name:"rk", Value:"true"}, libfido2.Option{Name:"up", Value:"true"}, libfido2.Option{Name:"plat", Value:"false"}, libfido2.Option{Name:"clientPin", Value:"false"}, libfido2.Option{Name:"credentialMgmtPreview", Value:"true"}}, Protocols:[]uint8{0x2, 0x1}} webauthncli/fido2.go:860
2024-08-08T16:46:14-03:00 DEBU             FIDO2: Device ioreg://4294988912 filtered, err=device already holds a registered credential webauthncli/fido2.go:869
2024-08-08T16:46:14-03:00 DEBU             FIDO2: Device ioreg://4294970869: filtered due to presence of excluded credential webauthncli/fido2.go:529
2024-08-08T16:46:14-03:00 DEBU             FIDO2: Device ioreg://4294970869 filtered, err=device already holds a registered credential webauthncli/fido2.go:869
2024-08-08T16:46:17-03:00 DEBU             FIDO2: Cancel device ioreg://4294970869, err=failed to cancel: tx error webauthncli/fido2.go:826
2024-08-08T16:46:17-03:00 DEBU             FIDO2: Close device ioreg://4294988912, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:46:17-03:00 DEBU             FIDO2: Cancel device ioreg://4294988912, err=<nil> webauthncli/fido2.go:826
2024-08-08T16:46:17-03:00 DEBU             FIDO2: Device touch status error: touch status: keep alive cancel webauthncli/fido2.go:1049
2024-08-08T16:46:17-03:00 DEBU             FIDO2: Close device ioreg://4294970869, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:46:17-03:00 DEBU             FIDO2: Device goroutines exited cleanly webauthncli/fido2.go:688
```

login (only authenticator):

```
2024-08-08T16:52:46-03:00 DEBU             FIDO2: Using libfido2 for assertion webauthncli/api.go:183
2024-08-08T16:52:46-03:00 DEBU             FIDO2: assertion: passwordless=false, uv=false, 1 allowed credentials webauthncli/fido2.go:168
Tap any security key
2024-08-08T16:52:46-03:00 DEBU             FIDO2: Device ioreg://4294988912: not a FIDO2 device webauthncli/fido2.go:862
2024-08-08T16:52:46-03:00 DEBU             FIDO2: Device ioreg://4294970869: info &libfido2.DeviceInfo{Versions:[]string{"U2F_V2", "FIDO_2_0", "FIDO_2_1_PRE"}, Extensions:[]string{"credProtect", "hmac-secret"}, AAGUID:[]uint8{0xee, 0x88, 0x28, 0x79, 0x72, 0x1c, 0x49, 0x13, 0x97, 0x75, 0x3d, 0xfc, 0xce, 0x97, 0x7, 0x2a}, Options:[]libfido2.Option{libfido2.Option{Name:"rk", Value:"true"}, libfido2.Option{Name:"up", Value:"true"}, libfido2.Option{Name:"plat", Value:"false"}, libfido2.Option{Name:"clientPin", Value:"false"}, libfido2.Option{Name:"credentialMgmtPreview", Value:"true"}}, Protocols:[]uint8{0x2, 0x1}} webauthncli/fido2.go:860
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Got 1 assertions webauthncli/fido2.go:251
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Authenticated: credential ID (b64) = e598-1ehGzWQY9liUvjYsaTDFKcClVlsIgT8VP4zkP9qAnkxWYKA7PcNVeToMZUuR1xaQa9HiPglEqJZcKnKrg, user ID (hex) = , user name = "" webauthncli/fido2.go:259
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Device ioreg://4294988912: callback returned, requiresPIN=false, err=<nil> webauthncli/fido2.go:883
Detected security key tap
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Cancel device ioreg://4294970869, err=<nil> webauthncli/fido2.go:826
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Close device ioreg://4294988912, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Cancel device ioreg://4294988912, err=<nil> webauthncli/fido2.go:826
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Device ioreg://4294970869: callback returned, requiresPIN=false, err=failed to get assertion: keep alive cancel webauthncli/fido2.go:883
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Close device ioreg://4294970869, err=<nil> webauthncli/fido2.go:842
2024-08-08T16:52:48-03:00 DEBU             FIDO2: Device goroutines exited cleanly webauthncli/fido2.go:688
```

**Comment by codingllama** [comment]
@ravicious, would you mind cloning this branch and running a final local test to make sure the fix is solid?

**Comment by codingllama** [comment]
Re backports: I've no plans to backport this as of now. I don't like changing this logic much, so I would rather only do that after v17 testplans (or if someone else raises the issue).

File: lib/auth/webauthncli/fido2.go [hunk 1]
```diff
@@ -312,13 +313,72 @@ func usesAppID(dev FIDODevice, info *deviceInfo, ccdHash []byte, allowedCreds []
...
+	ccdHash []byte,
+	allowedCreds [][]byte,
+) ([]byte, bool) {
+	const pin = ""
+	opts := &libfido2.AssertionOpts{
+		UP: libfido2.False,
+	}
+	for _, cred := range allowedCreds {
+		_, err := dev.Assertion(rpID, ccdHash, [][]byte{cred}, pin, opts)
+		if err == nil || (!info.fido2 && errors.Is(err, libfido2.ErrUserPresenceRequired)) {
```
Comment (smallinsky) [inline lib/auth/webauthncli/fido2.go:375]: “Could you add a comment for the reader explaining why non-FIDO2 devices don’t require user presence?

**Review by smallinsky — APPROVED** [review]

File: lib/auth/webauthncli/fido2.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/webauthncli/fido2.go:375]: It's a device/protocol difference. I'll see if I can find a link to substantiate.

**Comment by codingllama** [comment]
Friendly ping @gzdunek @ravicious ?

**Review by ravicious — APPROVED** [review]
I tested this locally and these changes let me log in with the old key.

File: lib/auth/webauthncli/fido2_test.go [hunk 2]
```diff
@@ -1954,6 +1954,102 @@ func TestFIDO2Register_u2fExcludedCredentials(t *testing.T) {
...
+	}
+
+	// Sanity check: authenticator errors in the presence of unknown credentials.
+	u2fDev.open()
+	_, err := u2fDev.Assertion(
+		"example.com",
+		[]byte(`55cde2973243a946b85a477d2e164a35d2e4f3daaeb11ac5e9a1c4cf3297033e`), // clientDataHash
+		[][]byte{
+			u2fDev.credentialID(),
+			[]byte("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/webauthncli/fido2_test.go:2026]: Done.

File: lib/auth/webauthncli/fido2.go (hunk 1 above)
Comment (codingllama) [reply lib/auth/webauthncli/fido2.go:378]: Added a comment and linked to libfido2 sources. I vaguely remember this mentioned in docs, but I couldn't find it in them.

- https://github.com/Yubico/libfido2/blob/03c18d396eb209a42bbf62f5f4415203cba2fc50/src/u2f.c#L787-L791

**Comment by codingllama** [comment]
Thanks everyone!

**Comment by codingllama** [comment]
FYI, I've moved the `errorOnUnknownCredential` after the tap check, as this better matches the behavior described by Rafal. (929de20)

**Pull request merged**